sync/src/types/header_map/kernel_lru.rs

```diff
@@ -0,0 +1,275 @@
+use std::path;
+
+#[cfg(feature = "stats")]
+use ckb_logger::trace;
+use ckb_types::packed::Byte32;
+
+use super::{KeyValueBackend, KeyValueMemory};
+use crate::types::HeaderView;
+
+pub(crate) struct HeaderMapLruKernel<Backend>
+where
+    Backend: KeyValueBackend,
+{
+    primary: KeyValueMemory<Byte32, HeaderView>,
+    backend: Backend,
+    // Configuration
+    primary_limit: usize,
+    backend_close_threshold: usize,
+    // Statistics
+    #[cfg(feature = "stats")]
+    stats: HeaderMapLruKernelStats,
+}
+
+#[cfg(feature = "stats")]
+#[derive(Default)]
+struct HeaderMapLruKernelStats {
+    frequency: usize,
+
+    trace_progress: usize,
+
+    primary_contain: usize,
+    primary_select: usize,
+    primary_insert: usize,
+    primary_delete: usize,
+
+    backend_contain: usize,
+    backend_insert: usize,
+    backend_delete: usize,
+}
+
+impl<Backend> HeaderMapLruKernel<Backend>
+where
+    Backend: KeyValueBackend,
+{
+    pub(crate) fn new<P>(
+        tmpdir: Option<P>,
+        primary_limit: usize,
+        backend_close_threshold: usize,
+    ) -> Self
+    where
+        P: AsRef<path::Path>,
+    {
+        let primary = Default::default();
+        let backend = Backend::new(tmpdir);
+
+        #[cfg(not(feature = "stats"))]
+        {
+            Self {
+                primary,
+                backend,
+                primary_limit,
+                backend_close_threshold,
+            }
+        }
+
+        #[cfg(feature = "stats")]
+        {
+            Self {
+                primary,
+                backend,
+                primary_limit,
+                backend_close_threshold,
+                stats: HeaderMapLruKernelStats::new(50_000),
+            }
+        }
+    }
+
+    pub(crate) fn has(&mut self, hash: &Byte32) -> bool {
```

Comment:
suggest to follow naming convention
```
has => contains_key / add => insert / rm => remove
```

Reply:
> suggest to follow naming convention
> 
> ```
> has => contains_key / add => insert / rm => remove
> ```

To avoid typo: use `map.inner.insert` instead of `map.insert`.
If they have different names, if we use the wrong method, it's a compile error.

```rust
fn insert(&mut self, value: HeaderView) -> Option<HeaderView>;
pub(crate) fn add(&mut self, view: HeaderView) -> Option<HeaderView>;
```

I could accept your suggestion only if these structs in different crates and we don't have chance to make mistakes.

Reply:
DONE and CI passed!